README

Gmail Backup Database

NAME
	gmdb -- manage local gmail message database

SYNOPSIS
	gmdb command [options] ...
	gmdb command -h
	gmdb -h

DESCRIPTION
	The gmdb client allows you backup all mail stored in a Gmail or Google Apps
	account to the local storage. The backup is performed using IMAP4 protocol,
	with some Gmail-specific extensions, which preserves the structure (labels)
	of all messages. The user has the option of selecting which messages are
	backed up by constructing a search query using the same syntax as the web
	interface search (see 'backup' command help for more information).

	By default, deleted messages are kept in the database for 1 year, but you
	may change this setting to any other value, including forever (see 'config'
	command). Messages may be restored either to Gmail servers (to the same or
	different account) or to any other IMAP4 server. The restoration can be
	performed for the entire account or for a limited number of messages (see
	'restore' command).

	The client relies heavily on the 'X-GM-EXT1' IMAP4 extensions, which are
	available on Gmail IMAP4 servers. As a result, the client is currently not
	able to back up mail from any other IMAP4 server, but this functionality may
	be added in a future release. Contacts, calendar entries, and chat logs are
	also not currently saved.

CONFIGURATION FILE
	An optional configuration file may be used to customize program behavior.
	The configuration file is required when authenticating to a Google Apps
	domain using OAuth (see next section). The default search path is:

	  (All)     ./gmdb.conf
	  (Windows) %USERPROFILE%\gmdb.conf
	  (Unix)    %HOME/.gmdb
	  (Unix)    PREFIX/etc/gmdb.conf  (where PREFIX is Python's sys.prefix)
	  (Unix)    /etc/gmdb.conf

	The exact path may also be specified using the --conf (-c) option. The
	configuration file uses standard Python syntax. The default settings are
	listed in _gmdb/conf.py. Here is an example configuration:

	  ROOT_DIR     = '/backup/gmail'             # Where backups are stored
	  OAUTH_KEY    = 'example.org'               # Google Apps domain name
	  OAUTH_SECRET = 'kmEgIBJL8ox3AKtTSL4TYzQU'  # OAuth consumer secret

AUTHENTICATION
	IMAP authentication may be performed using either plaintext password or
	2-legged OAuth mechanisms. The latter is intended for use with Google Apps
	domains and provides a way for administrators to back up all domain accounts
	without requiring individual user passwords.

	The plaintext password authentication should be used for individual Gmail
	accounts. Gmail requires SSL when connecting to their servers, so your
	password is encrypted before being sent over the wire. Even so, you should
	enable "2-step verification" for your account and use an application-
	specific password to avoid storing the real password in a file on your
	computer. See the following link for more information:

	  http://www.google.com/support/accounts/bin/static.py?page=guide.cs&
	  guide=1056283

	The preferred way to specify a password is through the --passwd-file (-P)
	option, which will read the password from a file, the standard input, or the
	current terminal (depending on what is connected to stdin). This is slightly
	more secure than specifying the password on the command line with the -p
	option, which is visible in the process list and is saved to your command
	history. You may also opt to enter the password manually for each
	backup/restore operation, which is the most secure, but slightly
	inconvenient method.

EXIT STATUS
	The program exists with a status of 0 on success, 1 on error, and 2 on an
	error in command line arguments.

STANDARDS
	Below is a list of informational URLs that are relevant to the client
	implementation. A more detailed list of IMAP-specific RFCs is at the top of
	_gmdb\imaplib2.py.

	* http://tools.ietf.org/html/rfc3501 - IMAP VERSION 4rev1
	* http://tools.ietf.org/html/rfc5849 - The OAuth 1.0 Protocol
	* http://code.google.com/apis/accounts/docs/OAuth.html - OAuth for Web Apps
	* http://code.google.com/apis/gmail/oauth/protocol.html - Gmail IMAP OAuth
	* http://code.google.com/apis/gmail/imap/ - Gmail IMAP Extensions

BUGS
	The 'muted' message status is not backed up. As far as I can tell, there is
	no way to retrieve it from the server.

	The 'starred' attribute is not converted to 'flagged' when restoring
	messages to Microsoft Exchange server.

	Creating a label called '\Important' and adding messages to it (which is
	not the same as marking messages 'important'), will cause those messages to
	be restored as 'important' rather than be labeled '\Important' (confused
	yet?). In other words, Gmail adds a special flag to each message when you
	mark it as important via the web interface. However, it doesn't prevent you
	from creating a label by the same name as the flag, which happens to be
	'\Important'. During a restore operation, the server can't tell whether
	'\Important' refers to the flag or a separate label, so it assumes the
	former.

	Running the source distribution on Windows requires an updated SQLite DLL
	in order for the 'index' command to work. The default DLL distributed with
	Python 3.2 does not have the FTS module compiled in. The solution is to
	download sqlite3.dll from the official SQLite website and replace the DLL
	stored in your Python32\DLLs directory.

	(Inconvenience) The restore operation is significantly slower than backup.
	This due to the design of IMAP4 protocol. During backup, the client can scan
	the attributes of 1000 messages (using the default configuration) and
	download all new messages in that range with just two commands. Restore, on
	the other hand, requires three command round-trips for each message. A
	number of extensions have been developed to improve message upload
	performance (RFC-2088 and RFC-3502), but Gmail servers do not support them.

	Pressing <Ctrl-C> at the 'Password:' prompt does not cause the program to
	exit immediately; you still need to press <Enter>. This a bug in Python 3.2,
	see issue #11236.

	The indexing operation may fail to parse some "unusual" (read: malformed)
	messages. If you encounter "Failed to parse message" errors, please send me
	the output of the following command (assuming that the message does not
	contain any sensitive information):

	  $ gmdb view -d <digest> <account>

AUTHOR
	Maxim Khitrov <max@mxcrypt.com>


CHANGES

gmdb v0.7.1                                                           2011-??-??

  *) Bugfix: challenge-response authentication was not properly encoding client
     response.

gmdb v0.7.0                                                           2011-07-25

  *) Feature: new 'archive' command for creating a single compressed file
     containing all messages in the database.

  *) Feature: the 'view' command now accepts a '-d <digest>' flag to view
     individual message content.

  *) Bugfix: an exception was raised after a new mailbox was successfully
     created during a restore operation to non-Gmail server.

  *) Bugfix: common IMAP queue processing could raise an exception if one of the
     responses did not contain any data.

  *) Bugfix: the 'shell' command raised an exception when the <Enter> key was
     pressed at an empty prompt.

  *) Bugfix: imaplib2 no longer calls shutdown on a closed socket after SSL
     certificate verification failure.

  *) Bugfix: message parsing failed when the Content-Type header wasn't
     separated by a CRLF from Content-Transfer-Encoding.

  *) Bugfix: message parsing failed if some of the headers contained unknown
     8-bit characters.
     Reported by Jason Smith.

  *) Bugfix: SIGINFO not available on Linux.
     Reported by Jason Smith.

  *) Selected 2-clause Simplified BSD License for public distribution


gmdb v0.6.0                                                           2011-07-20

  *) First internal release.


_gmdb/__init__.py

#
# Written by Maxim Khitrov (July 2011)
#

import sys
if sys.version_info < (3, 2):
	sys.exit('Python 3.2+ required')

try:
	import argparse
	import getpass
	import io
	import locale
	import logging
	import os
	import sqlite3
	import ssl
	import textwrap
except ImportError as exc:
	sys.exit(exc)

# Require SQLite 3.6.19+ for foreign key support
if sqlite3.sqlite_version_info < (3, 6, 19):
	sys.exit('Insufficient sqlite3 version (3.6.19+ required)')

__author__  = 'Maxim Khitrov'
__email__   = 'max@mxcrypt.com'
__version__ = '0.7.0'
__date__    = '2011-07-25'

#
###  I/O encoding  #############################################################
#

def switch_encoding(stream, encoding='UTF-8'):
	mode = getattr(stream, 'mode', None)
	lbuf = stream.line_buffering
	wrap = io.TextIOWrapper(stream.buffer, encoding, line_buffering=lbuf)
	if mode is not None:
		wrap.mode = mode
	return wrap

# Force stdout and stderr to use UTF-8 encoding
sys.stdout = switch_encoding(sys.stdout)
sys.stderr = switch_encoding(sys.stderr)

# Force open() to use UTF-8 encoding by default
locale.getpreferredencoding = lambda do_setlocale=True: 'UTF-8'

#
###  Logging  ##################################################################
#

logging.basicConfig(
	style   = '{',
	format  = '{asctime}  {levelname:8}  {message}',
	datefmt = '%Y-%m-%d %H:%M:%S',
	level   = logging.INFO
)

class Logger(logging.Logger):
	null   = lambda *args, **kwargs: None
	debug1 = null
	debug2 = null
	debug3 = null

class LogRecord(logging.LogRecord):
	def getMessage(self):
		msg = str(self.msg)
		if isinstance(self.args, dict):
			return msg.format(self.args)
		return msg.format(*self.args) if self.args else msg

logging.setLoggerClass(Logger)
logging.setLogRecordFactory(LogRecord)
logging.captureWarnings(True)

log = logging.getLogger('gmdb')

#
###  Main  #####################################################################
#

def main_wrapper():
	try:
		sys.exit(main(sys.argv))
	except KeyboardInterrupt:
		pass
	except Exception:
		log.exception('Unhandled error')
	sys.exit(1)

def main(argv):
	args = parse_args(argv[1:])
	verbose_output(args.verbose)

	# Load password from a file or the terminal
	if getattr(args, 'passwd_file', None):
		read_passwd(args)

	# Load configuration and copy command-line arguments into the conf module
	args.conf = load_conf(args.conf)
	for k, v in vars(args).items():
		setattr(conf, k, v)

	log.debug('Arguments: {}', args)
	cmd = Command(conf.command)
	return cmd()

def parse_args(args):
	argp = argparse.ArgumentParser(description='Gmail Backup Database')
	argp.add_argument('--version', action='version',
		version='%(prog)s ' + __version__)

	# Common options
	common = argparse.ArgumentParser(add_help=False)
	common.add_argument('-c', '--conf', metavar='FILE',
		type=argparse.FileType(), help='configuration file path')
	common.add_argument('-v', dest='verbose', action='count', default=0,
		help='enable verbose output (may be repeated up to 5 times)')

	# IMAP options
	imap = argparse.ArgumentParser(add_help=False)
	imap.add_argument('-s', '--server', metavar='HOST[:PORT]',
		default='imap.gmail.com:993',
		help='server hostname and port [imap.gmail.com:993]')
	imap.add_argument('-t', '--timeout', metavar='SEC', type=int, default=60,
		help='network socket timeout in seconds [60]')
	imap.add_argument('--ssl', action='store_true',
		help='use ssl to connect (on by default when port is 993)')
	imap.add_argument('-V', '--no-verify', action='store_true',
		help="do not verify server's SSL certificate")
	imap.add_argument('-z', '--compress', action='count', default=0,
		help='use link compression (repeat up to 3 times for level={3,6,9})')
	passwd = imap.add_mutually_exclusive_group()
	passwd.add_argument('-p', '--passwd',
		help='account password for plaintext authentication')
	passwd.add_argument('-P', '--passwd-file', metavar='FILE',
		type=argparse.FileType(), help='file containing account password')

	# DB options
	db = argparse.ArgumentParser(add_help=False)
	db.add_argument('-b', '--db-backup', action='store_true',
		help='copy sqlite database file before opening it')

	# Command-specific options
	subp = argp.add_subparsers(title='commands', dest='command')
	for name, cls in sorted(Command.registry.items()):
		parents = [common]
		if cls.uses_imap:
			parents.append(imap)
		if cls.uses_db:
			parents.append(db)
		first, rest = doc_format(cls.__doc__)
		cls.build_args(subp.add_parser(name,
			help            = first.lower().rstrip('.'),
			description     = '{}\n\n{}'.format(first, rest),
			formatter_class = argparse.RawDescriptionHelpFormatter,
			parents         = parents
		))

	return argp.parse_args(args)

def doc_format(doc):
	lines = doc.strip().splitlines()
	first = lines[0].rstrip()
	rest  = textwrap.dedent('\n'.join(lines[2:]))
	return (first, rest)

def verbose_output(level):
	if level:
		log.setLevel(logging.DEBUG)
		imaplib2.debug_level(level)
	for i, method in enumerate(('debug1', 'debug2', 'debug3'), 1):
		target = Logger.debug if i <= level else Logger.null
		setattr(Logger, method, target)

def read_passwd(args):
	fd = args.passwd_file
	args.passwd_file = fd.name
	if fd.isatty():
		args.passwd = getpass.getpass()
		if '\x03' in args.passwd:
			raise KeyboardInterrupt  # Bug fix for python 3.2; see issue 11236
	else:
		args.passwd = fd.readline().rstrip('\r\n')
		if not is_stdin(fd):
			fd.close()

def load_conf(fd=None):
	if fd is None:
		for file in conf.CONF_PATH:
			if os.path.isfile(file):
				fd = open(file)
				break
		else:
			return None
	try:
		ns   = {}
		name = fd.name if is_stdin(fd) else os.path.realpath(fd.name)
		log.debug('Loading configuration file: {}', name)
		exec(fd.read(), ns)
		for k, v in ns.items():
			if hasattr(conf, k) and k.upper() == k:
				setattr(conf, k, v)
		return name
	finally:
		if not is_stdin(fd):
			fd.close()

#
###  GMDB imports  #############################################################
#

from .     import conf, imaplib2
from .imap import IMAP4Control
from .db   import DBControl
from .cmd  import Command
from .util import is_stdin
